File: app/src/main/java/com/grassehh/app/configuration/TracingConfiguration.java
package com.grassehh.app.configuration;

import io.micrometer.context.ContextSnapshot.Scope;
import io.micrometer.context.ContextSnapshotFactory;
import io.micrometer.observation.ObservationRegistry;
import io.micrometer.observation.contextpropagation.ObservationThreadLocalAccessor;
import io.netty.channel.ChannelDuplexHandler;
import io.netty.channel.ChannelHandlerContext;
import io.netty.channel.ChannelPromise;
import jakarta.annotation.PostConstruct;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import reactor.core.publisher.Hooks;
import reactor.netty.Metrics;

@Configuration
public class TracingConfiguration {
    private ObservationRegistry observationRegistry;

    public TracingConfiguration(ObservationRegistry observationRegistry) {
        this.observationRegistry = observationRegistry;
    }

    @Bean
    public ContextSnapshotFactory contextSnapshotFactory() {
        return ContextSnapshotFactory.builder().build();
    }

    @PostConstruct
    public void postConstruct() {
        Hooks.enableAutomaticContextPropagation();
//        ContextRegistry.getInstance().registerThreadLocalAccessor(ObservationAwareSpanThreadLocalAccessor(tracer));
        ObservationThreadLocalAccessor.getInstance().setObservationRegistry(observationRegistry);
        Metrics.observationRegistry(observationRegistry);
    }
}

class TracingChannelDuplexHandler extends ChannelDuplexHandler {
    private ChannelDuplexHandler delegate;
    private ContextSnapshotFactory contextSnapshotFactory;

    public TracingChannelDuplexHandler(ChannelDuplexHandler delegate, ContextSnapshotFactory contextSnapshotFactory) {
        this.delegate = delegate;
        this.contextSnapshotFactory = contextSnapshotFactory;
    }

    @Override
    public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
        try (Scope scope = contextSnapshotFactory.setThreadLocalsFrom(ctx.channel())) {
            delegate.channelRead(ctx, msg);
        }
    }

    @Override
    public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) throws Exception {
        try (Scope scope = contextSnapshotFactory.setThreadLocalsFrom(ctx.channel())) {
            delegate.write(ctx, msg, promise);
        }
    }
}

File: app/src/main/kotlin/com/grassehh/app/configuration/TracingConfiguration.kt
package com.grassehh.app.configuration

import io.micrometer.context.ContextRegistry
import io.micrometer.context.ContextSnapshotFactory
import io.micrometer.core.instrument.kotlin.asContextElement
import io.micrometer.observation.ObservationRegistry
import io.micrometer.observation.contextpropagation.ObservationThreadLocalAccessor
import io.micrometer.tracing.Tracer
import io.micrometer.tracing.contextpropagation.ObservationAwareSpanThreadLocalAccessor
import io.netty.channel.ChannelDuplexHandler
import io.netty.channel.ChannelHandlerContext
import io.netty.channel.ChannelPromise
import jakarta.annotation.PostConstruct
import kotlinx.coroutines.withContext
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.web.server.CoWebFilter
import org.springframework.web.server.CoWebFilterChain
import org.springframework.web.server.ServerWebExchange
import reactor.core.publisher.Hooks
import reactor.netty.Metrics

@Configuration
class TracingConfiguration(private val observationRegistry: ObservationRegistry, private val tracer: Tracer) {
    @Bean
    fun coroutineWebFilter() = object : CoWebFilter() {
        override suspend fun filter(exchange: ServerWebExchange, chain: CoWebFilterChain) =
            withContext(observationRegistry.asContextElement()) {
                chain.filter(exchange)
            }
    }

    @Bean
    fun contextSnapshotFactory() = ContextSnapshotFactory.builder().build()

    @PostConstruct
    fun postConstruct() {
        Hooks.enableAutomaticContextPropagation()
        ContextRegistry.getInstance().registerThreadLocalAccessor(ObservationAwareSpanThreadLocalAccessor(tracer));
        ObservationThreadLocalAccessor.getInstance().observationRegistry = observationRegistry
        Metrics.observationRegistry(observationRegistry)
    }
}

class TracingChannelDuplexHandler(
    private val delegate: ChannelDuplexHandler,
    private val contextSnapshotFactory: ContextSnapshotFactory
) : ChannelDuplexHandler() {
    override fun channelRead(ctx: ChannelHandlerContext, msg: Any) {
        contextSnapshotFactory.setThreadLocalsFrom<String>(ctx.channel()).use {
            delegate.channelRead(ctx, msg)
        }
    }

    override fun write(ctx: ChannelHandlerContext, msg: Any, promise: ChannelPromise?) {
        contextSnapshotFactory.setThreadLocalsFrom<String>(ctx.channel()).use {
            delegate.write(ctx, msg, promise)
        }
    }
}